### DISCUSSION TITLE
Welcome to frads Discussions!

### DISCUSSION THREAD

**Discussion by taoning** [LBNL-ETA/frads#3] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by adalbertogc1** [comment]
Hello Frads group, 
My name is Adal and I work as a Building Physics Engineer. I am constantly looking for tools for evaluating the effectiveness of blind control algorithms. I look forward to exchanging ideas and hearing about how you use Frads in your projects.

**Comment by koepferl** [comment]
Hi, 

I am currently working through this tutorial.

https://lbnl-eta.github.io/frads/how-to/guide_rad1/

(base) ...@... ~ % gen room 12 14 3 \ 
    -w 0.4 1 2.5 1.8 \
    -w 3.3 1 2.5 1.8 \
    -w 6.2 1 2.5 1.8 \
    -w 9.1 1 2.5 1.8 \   
    -t 0.1 -n aroom

does not work. It returns me the following error:

usage: gen room [-h] [-w start_x start_z width height] [-n NAME] [-t Facade thickness] [-r ROTATE] width depth flrflr flrclg
gen room: error: the following arguments are required: flrclg

Just to be sure, I currently have version:

In [1]: import frads

In [2]: print(frads.version)
0.2.10

However the installation of the below did not quite work, because a file was missing:

pip install git+https://github.com/taoning/pyenergyplus_wheels
Collecting git+https://github.com/taoning/pyenergyplus_wheels
Cloning https://github.com/taoning/pyenergyplus_wheels to /private/var/folders/c2/tthc2_6j741c3rvx9vpjppwh0000gn/T/pip-req-build-ihxnqfnq
Running command git clone --filter=blob:none --quiet https://github.com/taoning/pyenergyplus_wheels /private/var/folders/c2/tthc2_6j741c3rvx9vpjppwh0000gn/T/pip-req-build-ihxnqfnq
Resolved https://github.com/taoning/pyenergyplus_wheels to commit 664456bc5428ad50ace2c2bc729bbc486afa0435
Preparing metadata (setup.py) ... done
Collecting pyenergyplus@ https://github.com/taoning/pyenergyplus/raw/main/wheels/darwin/pyenergyplus-23.1.0-py3-none-macosx_10_13_x86_64.whl
ERROR: HTTP error 404 while getting https://github.com/taoning/pyenergyplus/raw/main/wheels/darwin/pyenergyplus-23.1.0-py3-none-macosx_10_13_x86_64.whl
ERROR: Could not install requirement pyenergyplus@ https://github.com/taoning/pyenergyplus/raw/main/wheels/darwin/pyenergyplus-23.1.0-py3-none-macosx_10_13_x86_64.whl from https://github.com/taoning/pyenergyplus/raw/main/wheels/darwin/pyenergyplus-23.1.0-py3-none-macosx_10_13_x86_64.whl (from pyenergyplus-wheels==0.1.0) because of HTTP error 404 Client Error: Not Found for url: https://github.com/taoning/pyenergyplus/raw/main/wheels/darwin/pyenergyplus-23.1.0-py3-none-macosx_10_13_x86_64.whl for URL https://github.com/taoning/pyenergyplus/raw/main/wheels/darwin/pyenergyplus-23.1.0-py3-none-macosx_10_13_x86_64.whl

**Reply by koepferl** [reply]
changing it to gen room 12 14 3 0 \ did work.

**Reply by koepferl** [reply]
However, it does not seem to find the next module within python

from frads import matrix
from frads import parsers
from frads import sky
from frads import utils
from frads.types import WeaData
from frads.types import WeaMetaData

fpaths = ["Objects/materials_aroom.mat",
          "Objects/ceiling_aroom.rad",
          "Objects/wall_aroom.rad",
          "Objects/floor_aroom.rad",
          "Objects/window_00_aroom.rad",
          "Objects/window_01_aroom.rad",
          "Objects/window_02_aroom.rad",
          "Objects/window_03_aroom.rad",
         ]

room_octree = "aroom.oct"
matrix.oconv(*fpaths, room_octree)

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/var/folders/c2/tthc2_6j741c3rvx9vpjppwh0000gn/T/ipykernel_70799/2306175151.py in <module>
     17 
     18 room_octree = "aroom.oct"
---> 19 matrix.oconv(*fpaths, room_octree)

AttributeError: module 'frads.matrix' has no attribute 'oconv'


​

**Reply by koepferl** [reply]
<img width="552" alt="Screenshot 2024-02-18 at 21 24 01" src="https://github.com/LBNL-ETA/frads/assets/2587204/90953ea3-0da5-4ed2-8fbd-3b4ee345f60a">

BTW I fixed the other issues, there are small bugs in the tutorial regarding datetime

**Comment by koepferl** [comment]
However, then within python it does not seem to find matrix.oconv

`from frads import matrix
from frads import parsers
from frads import sky
from frads import utils
from frads.types import WeaData
from frads.types import WeaMetaData

fpaths = ["Objects/materials_aroom.mat",
          "Objects/ceiling_aroom.rad",
          "Objects/wall_aroom.rad",
          "Objects/floor_aroom.rad",
          "Objects/window_00_aroom.rad",
          "Objects/window_01_aroom.rad",
          "Objects/window_02_aroom.rad",
          "Objects/window_03_aroom.rad",
         ]

room_octree = "aroom.oct"
matrix.oconv(*fpaths, room_octree)`

`---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/var/folders/c2/tthc2_6j741c3rvx9vpjppwh0000gn/T/ipykernel_70799/2306175151.py in <module>
     17 
     18 room_octree = "aroom.oct"
---> 19 matrix.oconv(*fpaths, room_octree)

AttributeError: module 'frads.matrix' has no attribute 'oconv'
`

**Comment by taoning** [comment]
Hi @koepferl,  there are some mistakes in this tutorial,  we will fix the tutorial asap.

**Reply by koepferl** [reply]
Thank you ;)

**Reply by koepferl** [reply]
@taoning I looked through your fixes. I spotted a few things:

1) ``Objects/windows_0?_aroom.rad`` have an extra row of zeros.
2) ``gen_perez_sky`` there are a few issues with input, see the fix below

``date_time = datetime.datetime(2024, 12, 21, 12, 0)
wea_data = fr.WeaData(time=date_time, dni=800, dhi=100)


city = 'city'
county = 'county'
latitude = 37
longitude = 122
time_zone = 120
elevation = 0

wea_meta = fr.WeaMetaData(city, county, latitude, longitude, time_zone, elevation)

sky_descr = fr.gen_perez_sky(wea_data, wea_meta)``

3) ``pr.oconv`` gets me now this weird error

``room_sky_octree = f'aroom_37_122_1221_1200.oct'
#room_sky_octree = f'aroom_{latitude}_{longitude}_{month:02d}{day:02d}_{hour:02d}{minute:02d}.oct'

with open(room_sky_octree, "wb") as f:
    f.write(pr.oconv(stdin=sky_descr, octree=room_octree))``

``---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/var/folders/c2/tthc2_6j741c3rvx9vpjppwh0000gn/T/ipykernel_73642/1239290419.py in <module>
      3 
      4 with open(room_sky_octree, "wb") as f:
----> 5     f.write(pr.oconv(stdin=sky_descr, octree=room_octree))

~/opt/anaconda3/lib/python3.9/site-packages/pyradiance/anci.py in wrapper(*args, **kwargs)
     20     def wrapper(*args, **kwargs):
     21         try:
---> 22             result = func(*args, **kwargs)
     23         except CalledProcessError as e:
     24             errmsg = e.stderr or b""

~/opt/anaconda3/lib/python3.9/site-packages/pyradiance/ot.py in oconv(warning, stdin, frozen, octree, *paths)
     53             cmd.append("-")
     54         else:
---> 55             raise TypeError("stdin should be bytes.")
     56     return sp.run(cmd, input=stdin, stdout=sp.PIPE, check=True).stdout

TypeError: stdin should be bytes.``

**Reply by koepferl** [reply]
@taoning also I see no windows with ``objview Objects/*aroom.mat Objects/*aroom.rad`` now

**Reply by taoning** [reply]
Sorry it's not yet ready, will let you know when.  Also the some of the issues you are seeing are due to you having an older version of frads. So we need to sort that out first.

**Reply by koepferl** [reply]
Ok, sorry.  Thanks for your help

**Reply by koepferl** [reply]
Thanks for your help, in the end it was a weird issue due to my installation of the wrong anaconda version (for intel) rather than (apple M1) which did not cause any issue in python, but in the connection with radiance the issue appeared. thanks.

**Comment by koepferl** [comment]
Hi, thanks for your help. I am looking for daylight factor simulation of a simple room. Is the daylight factor part of frads? Or can you point me towards any other (hopefully python based) solution. 

Cheers, Christine

P.S. I am teaching programming for building technology engineers and I want to introduce them a little to daylight simulation. My background is maybe unusual I am a physicists and performed radiative transfer simulation in astrophysical environments but always in connection with python which I also teach. So I might have unusual questions.

**Reply by taoning** [reply]
daylight factor calculation has not been our focus but we could put together something in the tutorial.

also, If you want to use energyplus for building energy simulation you can use frads as well.  We have built in energyplus object model and simulation utilities.  when you install frads, you install both radiance and energyplus, so you don't need to set it up separately.

**Reply by koepferl** [reply]
This sounds great, the daylight factor is quite important for the focus of the lecture, as it is part of some of the EUs regulations.

**Reply by TammieYu** [reply]
@koepferl Please see our updated guide on the daylight factor [here](https://lbnl-eta.github.io/frads/how-to/guide_rad7/) [https://lbnl-eta.github.io/frads/how-to/guide_rad7/].

**Reply by koepferl** [reply]
@TammieYu @taoning sorry I never replied, the semester started and I did not have time to give it a look. This looks wonderful. Great.

**Reply by koepferl** [reply]
@TammieYu @taoning Unfortunately it does not work. I have corrected a few typos 

1. datatime -> datetime
2. stdin = sky_descr -> stdin=sky_scene

I hope this was the correct interpretation. However, there is a bug I can not fix following when setting up the sensors. 

floor_primitives = fr.unpack_primitives("Objects/floor_aroom.rad")
print(floor_primitives[0])
floor_polygon = fr.parse_polygon(floor_primitives[0])
grid = fr.gen_grid(floor_polygon, 1, 0.75)

neutral_lambertian_0.2 polygon floor
0 
0
12 0.0 0.0 0.0 12.0 0.0 0.0 12.0 14.0 0.0 0.0 14.0 0.0

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/var/folders/c2/tthc2_6j741c3rvx9vpjppwh0000gn/T/ipykernel_75884/3123292342.py in <module>
      1 floor_primitives = fr.unpack_primitives("Objects/floor_aroom.rad")
      2 print(floor_primitives[0])
----> 3 floor_polygon = fr.parse_polygon(floor_primitives[0])
      4 grid = fr.gen_grid(floor_polygon, 1, 0.75)

~/opt/anaconda3/lib/python3.9/site-packages/frads/parsers.py in parse_polygon(real_args)
    277     """
    278     coords = real_args[1:]
--> 279     arg_cnt = int(real_args[0])
    280     vertices = [geom.Vector(*coords[i : i + 3]) for i in range(0, arg_cnt, 3)]
    281     return geom.Polygon(vertices)

ValueError: invalid literal for int() with base 10: 'neutral_lambertian_0.2'

**Comment by koepferl** [comment]
Hi, I experience a very interesting malfunction. This might be the reason will many of the tutorials are not working for me. 

When I check for the frads version in the terminal I get 1.2.5

(base) greif@maccmg ~ % ipython
Python 3.11.5 (main, Sep 11 2023, 08:31:25) [Clang 14.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.15.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import frads

In [2]: frads.__version__
Out[2]: '1.2.5'

When I check in the Jupyter Notebook:

import frads
frads.__version__
'0.2.10' 

Any ideas?

**Reply by koepferl** [reply]
I moved to script to avoid this problem vor now. I don't get the same result as you. All df values are around 20? 

The only thing I changed was the name of the sky passed to the octree. Did you maybe use a sky? 

Below is my code:

####################################
# How to calculate daylight factor?
####################################

#https://lbnl-eta.github.io/frads/how-to/guide_rad7/

# 0. Import the required classes and functions

import datetime
import frads as fr
import pyradiance as pr
import numpy as np

import os

#############
# build room
#############

# 1. Build a Radiance model

os.system('gen room 12 14 4 3 \
    -w 0.4 1 2.5 1.8 \
    -w 3.3 1 2.5 1.8 \
    -w 6.2 1 2.5 1.8 \
    -w 9.1 1 2.5 1.8 \
    -t 0.1 -n aroom')#

# all moved to Object directory

# objview Objects/*aroom.mat Objects/*aroom.rad 
# p 180


fpaths = ["Objects/materials_aroom.mat",
          "Objects/ceiling_aroom.rad",
          "Objects/wall_aroom.rad",
          "Objects/floor_aroom.rad",
          "Objects/window_00_aroom.rad",
          "Objects/window_01_aroom.rad",
          "Objects/window_02_aroom.rad",
          "Objects/window_03_aroom.rad",
]

room_octree = "aroom.oct"
with open(room_octree, 'wb') as f:
    f.write(pr.oconv(*fpaths))
    
##############
# light source
##############

# 2. Generate a CIE overcast sky

dif_hor_illum = 10000
dif_hor_ird = dif_hor_illum / 179

sky_func = pr.gensky(
    dt=datetime.datetime(2024, 12, 21, 12, 0),
    cloudy=True,
    horizontal_brightness=dif_hor_ird, # 
    )

print(sky_func)

sky_glow = "skyfunc glow sky_glow 0 0 4 1 1 1 0".encode()

sky = "sky_glow source sky 0 0 4 0 0 1 180".encode()

ground = "sky_glow source ground 0 0 4 0 0 -1 180".encode()

sky_scene = sky_func + b'\n' + sky_glow + b'\n' + sky + b'\n' + ground

# add sky scene to octree
room_sky_octree = f"aroom_37_122_1221_1200.oct"
with open(room_sky_octree, "wb") as f:
    f.write(pr.oconv(stdin=sky_scene, octree=room_octree))
    

# 3. Use raytracing to get the interior horizontal illuminance

##############
# Sensors
##############

# generate grid sensors based on the floor polygon
floor_primitives = fr.unpack_primitives("Objects/floor_aroom.rad")
floor_polygon = fr.parse_polygon(floor_primitives[0])
grid = fr.gen_grid(floor_polygon, 1, 0.75)

##############
# tracing rays
##############

# rtrace to get irradiance
option = ["-I+", "-ab", "5", "-ad", "512", "-aa", "0", "-lw", "0.001"]
rays = "\n".join([" ".join(map(str, row)) for row in grid])
ird_results = pr.rtrace(rays.encode(), room_sky_octree, params=option, header=False)

print(ird_results[:50])

##############
# Postprocessing
##############

# reformat the irradiance result

rows = ird_results.decode().strip().split("\n")
data = [
    [float(element) for element in row.split("\t") if element != ""]
    for row in rows
]
ird_array = np.array(data)

# convert irradiance to illuminance
illum = ird_array[:, 0] * 47.4 + ird_array[:, 1] * 119.9 + ird_array[:, 2] * 11.6 # 

##############
# daylight factor
##############

# 4. Compute the daylight factor
# daylight factor = interior illuminance / exterior illuminance * 100

df = illum / dif_hor_illum * 100

print('df min max', min(df), max(df))

##############
# plotting
##############

# import visualization packages

import matplotlib.pyplot as plt

fig, ax = plt.subplots()
x = [data[0] for data in grid]
y = [data[1] for data in grid]
_plot = ax.scatter(
    x,
    y,
    c=df,
    cmap="plasma",
    s=15,
    vmin=min(df),
    vmax=max(df),
    rasterized=True,
)
ax.set(
    xlabel="x position [m]",
    ylabel="y position [m]",
)

fig.colorbar(_plot, ax=ax, label="Daylight Factor [%]")
fig.tight_layout()
fig.savefig('df.png')

**Reply by taoning** [reply]
It's not immediately obviously to me.  What's your `grid` and `ird_results` variable.

**Reply by koepferl** [reply]
Hi, I followed this tutorial. https://lbnl-eta.github.io/frads/how-to/guide_rad7/

``ird_results = pr.rtrace(rays.encode(), room_sky_octree, params=option, header=False)``

``grid = fr.gen_grid(floor_polygon, 1, 0.75)``

**Reply by koepferl** [reply]
[How_to_calculate_daylight_factor.pdf](https://github.com/user-attachments/files/15793818/How_to_calculate_daylight_factor.pdf)

**Reply by koepferl** [reply]
I get back this: 
![df](https://github.com/LBNL-ETA/frads/assets/2587204/68df0992-3258-404d-80f7-11a9151ccc44)

**Reply by taoning** [reply]
Yes, can you print out those two variables?

**Comment by koepferl** [comment]
[Terminal Saved Output.txt](https://github.com/user-attachments/files/15798728/Terminal.Saved.Output.txt)

**Reply by koepferl** [reply]
Another question, how can I copy code here into GitHub chat that it does not blow up in the Markdowns for every comment starting with #

**Reply by taoning** [reply]
It looks like the grid was not generated properly.  They are underneath the building facing downwards.  We changed this some time ago, and forgot to update the tutorial.
Can you try this instead:
```
grid = fr.gen_grid(floor_polygon.flip(), 1, 0.75)
```

You can format the code using the code in the editor or encapsulate the code with triple apostrophe ```

**Reply by koepferl** [reply]
Thanks, a lot that works. I created a pretty nice plot of the evolution of the daylight factor in the room interior. Thanks for your help.

**Reply by koepferl** [reply]
Do you have an idea why it I have a totally different version of frads in the Jupyter notebook. There I have 0.XX on the system I have 1.XX

**Reply by taoning** [reply]
Sounds like your jupyter notebook is picking up a different virtual environment. Have you tried to run `!pip install -U frads` inside? Maybe it's also using an older version of python.

**Reply by koepferl** [reply]
'0.2.10' if I do the pip install -U in the notebook (3.9.13)
'1.2.5' in the shell (Python 3.11.5)

Weird, I need to also say that in February you helped me because of the introduction of the M1 chip in Apple devices the initial anaconda installation I installed was not for M1, but I thought this was resolved. Do you have any ideas?
![df3](https://github.com/LBNL-ETA/frads/assets/2587204/d2237437-5807-4c0f-8ee5-c57046034ae5)

I enclose an image which I managed to make from your tutorials and adapted a daylight factor map. 

Is there a way to increase the number of sensors in the grid?

**Reply by taoning** [reply]
I don't know why that's the case, but I am also not surprised given how difficult it is to manage python environment. I guess the only "solution" would be the construct a virtual environment for every project and installing fresh package in it.  But I also understand that's pain if you want to work inside a notebook.

Thanks for including the image. You can decrease the `spacing` value from you `gen_grid` call to get more sensors.

**Reply by koepferl** [reply]
I solved the problem by entire getting rid of python from my computer. I had two running installations. After reinstalling everything seems to work - fingers crossed.

**Comment by robertweitl** [comment]
### Post Process of eplusout.sql or eplusout.csv.
Is there any built-in method to postprocess the eplus results? The Documentation doesn't anything on this. Or do just panda-read the csv to a dataframe and keep on going on the df?

btw, I am Robert, a building energy modeller ;)

**Reply by taoning** [reply]
Hello Robert, the building energy modeler. 
There isn't right now, but should be.
We thought about building a EnergyPlusResult object that's built on the dataset concept in xarray, but haven't got around to do that.
Right now the solution would be to just load the csv file using something like pandas.
